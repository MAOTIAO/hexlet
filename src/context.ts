import shiki from 'shiki';
import { Config } from './config';
import { FormattedString, T } from './formattedString';

/**
 * Internal context object used to pass around config and config-derived
 * constants.
 */
export type Context = Config & {
    SPLIT_DIFFS: boolean;
    LINE_W