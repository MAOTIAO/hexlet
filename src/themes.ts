
import * as assert from 'assert';
import * as path from 'path';
import * as fs from 'fs';

const THEMES_DIR = path.resolve(__dirname, '..', 'themes');

/**
 * Colors are always specified as hex strings
 */
type Color = string;

type ColorModifier =
    | 'reset'
    | 'bold'
    | 'dim'
    | 'italic'
    | 'underline'
    | 'inverse'
    | 'hidden'
    | 'strikethrough'
    | 'visible';

type ColorDefinition = {
    color?: Color;
    backgroundColor?: Color;
    modifiers?: ColorModifier[];
};

export enum ThemeColorName {
    DEFAULT_COLOR = 'DEFAULT_COLOR',
    COMMIT_HEADER_LABEL_COLOR = 'COMMIT_HEADER_LABEL_COLOR',
    COMMIT_HEADER_COLOR = 'COMMIT_HEADER_COLOR',
    COMMIT_SHA_COLOR = 'COMMIT_SHA_COLOR',
    COMMIT_AUTHOR_COLOR = 'COMMIT_AUTHOR_COLOR',
    COMMIT_DATE_COLOR = 'COMMIT_DATE_COLOR',
    COMMIT_TITLE_COLOR = 'COMMIT_TITLE_COLOR',
    COMMIT_MESSAGE_COLOR = 'COMMIT_MESSAGE_COLOR',
    BORDER_COLOR = 'BORDER_COLOR',
    FILE_NAME_COLOR = 'FILE_NAME_COLOR',
    HUNK_HEADER_COLOR = 'HUNK_HEADER_COLOR',
    DELETED_WORD_COLOR = 'DELETED_WORD_COLOR',
    DELETED_LINE_COLOR = 'DELETED_LINE_COLOR',
    DELETED_LINE_NO_COLOR = 'DELETED_LINE_NO_COLOR',
    INSERTED_WORD_COLOR = 'INSERTED_WORD_COLOR',
    INSERTED_LINE_COLOR = 'INSERTED_LINE_COLOR',
    INSERTED_LINE_NO_COLOR = 'INSERTED_LINE_NO_COLOR',
    UNMODIFIED_LINE_COLOR = 'UNMODIFIED_LINE_COLOR',
    UNMODIFIED_LINE_NO_COLOR = 'UNMODIFIED_LINE_NO_COLOR',
    MISSING_LINE_COLOR = 'MISSING_LINE_COLOR',
}

export type ThemeDefinition = {
    SYNTAX_HIGHLIGHTING_THEME?: string;
} & {
    [key in ThemeColorName]: ColorDefinition;
};

type ColorRgba = {
    r: number;
    g: number;
    b: number;
    a: number;
};

export type ThemeColor = {
    color?: ColorRgba;
    backgroundColor?: ColorRgba;
    modifiers?: ColorModifier[];
};

/**
 * The hex string is of the format #rrggbb(aa)
 */
function hexToRgba(hex: string): ColorRgba {
    assert.ok(hex.length === 7 || hex.length === 9);

    let hexNo = parseInt(hex.slice(1), 16);

    let a = 255;
    if (hex.length === 9) {
        a = hexNo & 0xff;
        hexNo >>>= 8;
    }

    const b = hexNo & 0xff;
    hexNo >>>= 8;

    const g = hexNo & 0xff;
    hexNo >>>= 8;

    const r = hexNo & 0xff;

    return { r, g, b, a };
}

export function mergeColors(
    a?: ColorRgba,
    b?: ColorRgba
): ColorRgba | undefined {
    if (!b || b.a === 0) {
        return a;
    }
    if (!a) {
        return b;
    }
    const t = 1.0 - b.a / 255.0;
    return {
        r: b.r * (1 - t) + a.r * t,
        g: b.g * (1 - t) + a.g * t,
        b: b.b * (1 - t) + a.b * t,
        a: b.a * (1 - t) + a.a * t,
    };
}

function mergeModifiers(
    a: ColorModifier[] | undefined,
    b: ColorModifier[] | undefined
): ColorModifier[] | undefined {
    if (a && b) {
        return a.concat(b);
    }
    return a ?? b;
}

function mergeThemeColors(a: ThemeColor, b: ThemeColor): ThemeColor {
    return {
        color: mergeColors(a.color, b.color),
        backgroundColor: mergeColors(a.backgroundColor, b.backgroundColor),
        modifiers: mergeModifiers(a.modifiers, b.modifiers),
    };
}

export function reduceThemeColors(colors: ThemeColor[]): ThemeColor {
    // Colors are applied in reverse, which allows us to do apply specific
    // formatting (like syntax highlighting) first and apply more generic colors
    // (like line colors) last
    let themeColor: ThemeColor = {};
    for (let i = colors.length - 1; i >= 0; i--) {
        themeColor = mergeThemeColors(themeColor, colors[i]);