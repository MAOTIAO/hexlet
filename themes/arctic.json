{
    "SYNTAX_HIGHLIGHTING_THEME": "nord",
    "DEFAULT_COLOR": {
        "color": "#E5E9F0",
        "backgroundColor": "#2b2b2b"
    },
    "COMMIT_HEADER_COLOR": {
        "color": "#4C566A"
    },
    "COMMIT_HEADER_LABEL_COLOR": {
        "color": "#88C0D0"
    },
    "COMMIT_SHA_COLOR": {
        "color": "#BF616A"
    },
    "COMMIT_AUTHOR_COLOR": {
        "color": "#D08770"
    },
    "COMMIT_DATE_COLOR": {
        "color": "#EBCB8B"
    },
    "COMMIT_MESSAGE_COLOR": {
        "color": "#D8DEE9"
    },
    "COMMIT_TITLE_COLOR": {
        "color": "#ECEFF4"
    },
    "FILE_NAME_COLOR": {
        "color": "#EBCB8B"
    },
    "BORDER_COLOR": {
        "color": "#EBCB8B"
    },
    "HUNK_HEADER_COLOR": {
        "color": "#4C566A"
    },
    "DELETED_WORD_COLOR": {
        "color": "#EBCB8B",
        "backgroundColor": "#BF616A33"
    },
    "INSERTED_WORD_COLOR": {
        "color": "#EBCB8B",
        "backgroundColor": "#A3BE8C33"
    },
    "DELETED_LINE_NO_COLOR": {
        "color": "#BF616A"
    },
    "INSERTED_LINE_NO_COLOR": {
        "color": "#A3BE8C"
    },
    "UNMODIFIED_LINE_NO_COLOR": {
        "color": "#4C566A"
    },
    "DELETED_LINE_COLOR": {
        "color": "#BF616A",
        "backgroundColor": "#3a3035"
    },
    "INSERT