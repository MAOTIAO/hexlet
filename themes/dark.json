{
    "SYNTAX_HIGHLIGHTING_THEME": "dark-plus",
    "DEFAULT_COLOR": {
        "color": "#ffffff",
        "backgroundColor": "#2b2b2b"
    },
    "COMMIT_HEADER_COLOR": {
        "color": "#cccccc"
    },
    "COMMIT_HEADER_LABEL_COLOR": {
        "color": "#00000022"
    },
    "COMMIT_SHA_COLOR": {
        "color": "#00eeaa"
    },
    "COMMIT_AUTHOR_COLOR": {
        "color": "#00aaee"
    },
    "COMMIT_DATE_COLOR": {
        "color": "#cccccc"
    },
    "COMMIT_MESSAGE_COLOR": {
        "color": "#cccccc"
    },
    "COMMIT_TITLE_COLOR": {
        "modifiers": ["bold"]
    },
    "FILE_NAME_COLOR": {
        "color": "#ffdd99"
    },
    "BORDER_COLOR": {
        "color": "#ffdd9966",
        "modifiers": ["dim"]
    },
    "HUNK_HEADER_COLOR": {
        "modifiers": ["dim"]
    },
    "DELETED_WORD_COLOR": {
        "color": "#ffcccc",
        "backgroundColor": "#ff000033"
    },
    "INSERTED_WORD_COLOR": {
        "color": "#ccffcc",
        "backgroundColor": "#00ff00