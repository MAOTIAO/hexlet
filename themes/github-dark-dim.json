{
    "SYNTAX_HIGHLIGHTING_THEME": "github-dark",
    "DEFAULT_COLOR": {
        "color": "#768390",
        "backgroundColor": "#22282d"
    },
    "COMMIT_HEADER_COLOR": {
        "color": "#adbac7",
        "backgroundColor": "#22272e"
    },
    "COMMIT_HEADER_LABEL_COLOR": {},
    "COMMIT_