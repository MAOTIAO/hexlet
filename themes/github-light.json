{
    "SYNTAX_HIGHLIGHTING_THEME": "github-light",
    "DEFAULT_COLOR": {
        "color": "#6a737d",
        "backgroundColor": "#ffffff"
    },
    "COMMIT_HEADER_COLOR": {
        "color": "#24292e",
        "backgroundColor": "#f1f8fe"
    },
    "COMMIT_HEADER_LABEL_COLOR": {
        "color": "#586069"
    },
    "COMMIT_AUTHOR_COLOR": {
        "color": "#444d56"
    },
    "COMMIT_SHA_COLOR": {
        "color": "#444d56"
    },
    "COMMIT_DATE_COLOR": {
        "color": "#444d56"
    },
    "COMMIT_MESSAGE_COLOR": {
        "color": "#444d56",
        "backgroundColor": "#f1f8fe"
    },
    "COMMIT_TITLE_COLOR": {
        "color": "#05264c",
        "modifiers": ["bold"]
    },
    "BORDER_COLOR": {
        "color": "#e1e4e8"
    },
    "FILE_NAME_COLOR": {
        "color": "#05264c"
    },
    "HUNK_HEADER_COLOR": {
        "color": "#4c5154",
        "backgroundColor": "#f1f8ff"
    },
    "DELETED_WORD_COLOR": {
        "backgroundColor": "#fdb8c0"
    },
    "INSERTED_WORD_COLOR": {
        "backgroundColor": "#acf2bd"
    },
    "DELETED_LINE_NO_COLOR": {
        "color": "#c2a9ad",
        "backgroundColor": "#ffdce0"
    },
    "INSERTED_LINE_NO_COLOR": {
        "color": "#a5cfb0",
        "backgroundColor": "#cdffd8"
    },
    "UNMODIFIED_LINE_NO_COLOR": {
        "color": "#d1d5da"
    },
    "DELETED_LINE_COLOR": {
        "color": "#24292e",
        "backgroundColor": "#ffeef0"
    },
    "INSERTED_LINE_COLOR": {
        "color": "#24292e",
        "backgroundColor": "#e6ffed"
    },
    "UNMODIFIED_LINE_COLOR": {},
    "MISSING_LINE_COLOR": {
        "backgroundColor": "#fafbfc"
    }
}
