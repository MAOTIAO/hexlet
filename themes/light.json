
{
    "SYNTAX_HIGHLIGHTING_THEME": "light-plus",
    "DEFAULT_COLOR": {
        "color": "#666666",
        "backgroundColor": "#eeeeee"
    },
    "COMMIT_HEADER_COLOR": {
        "color": "#333333"
    },
    "COMMIT_HEADER_LABEL_COLOR": {},
    "COMMIT_SHA_COLOR": {
        "color": "#006600"
    },
    "COMMIT_AUTHOR_COLOR": {
        "color": "#663333"
    },
    "COMMIT_DATE_COLOR": {},
    "COMMIT_MESSAGE_COLOR": {
        "color": "#666666"
    },
    "COMMIT_TITLE_COLOR": {
        "color": "#333333"
    },
    "FILE_NAME_COLOR": {
        "color": "#336666"
    },
    "BORDER_COLOR": {
        "color": "#cccccc"
    },
    "HUNK_HEADER_COLOR": {
        "color": "#666666",
        "backgroundColor": "#e0e0e0"
    },
    "DELETED_WORD_COLOR": {
        "backgroundColor": "#eeaaaa"
    },
    "INSERTED_WORD_COLOR": {
        "backgroundColor": "#aaeeaa"
    },
    "DELETED_LINE_NO_COLOR": {
        "color": "#ff000022",
        "backgroundColor": "#00000011"
    },
    "INSERTED_LINE_NO_COLOR": {
        "color": "#00ff0022",
        "backgroundColor": "#00000011"
    },
    "UNMODIFIED_LINE_NO_COLOR": {
        "color": "#ffffff88",
        "backgroundColor": "#00000008"
    },
    "DELETED_LINE_COLOR": {
        "color": "#333333",
        "backgroundColor": "#eecccc"
    },
    "INSERTED_LINE_COLOR": {
        "color": "#333333",
        "backgroundColor": "#cceecc"
    },
    "UNMODIFIED_LINE_COLOR": {},
    "MISSING_LINE_COLOR": {
        "backgroundColor": "#e8e8e8"
    }
}