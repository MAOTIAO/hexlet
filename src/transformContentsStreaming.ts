import stream, { Readable, Writable } from 'stream';
import { Context } from './context';
import { iterlinesFromReadable } from './iterLinesFromReadable';
import { iterReplaceTabsWithSpaces } from './iterReplaceTabsWithSpaces';
import { iterSideBySideDiffs } from './iterS