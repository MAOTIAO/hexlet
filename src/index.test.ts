
import { Readable, Writable } from 'stream';
import { Config } from './config';
import { getContextForConfig } from './context';
import { ThemeColorName } from './themes';
import { transformContentsStreaming } from './transformContentsStreaming';

const TEST_THEME = Object.fromEntries(
    Object.keys(ThemeColorName).map((name) => [name, {}])
);

const replaceColoredText = () => (text: string) => text.replace(/./g, 'â–‘');

const TEST_CONFIG: Config = {
    // Provide a fake chalk implementation to make it easier to read snapshots
    CHALK: {
        // @ts-expect-error
        rgb: replaceColoredText,
        // @ts-expect-error
        bgRgb: replaceColoredText,
    },
    SCREEN_WIDTH: 120,
    MIN_LINE_WIDTH: 60,
    WRAP_LINES: false,
    HIGHLIGHT_LINE_CHANGES: false,
    ...TEST_THEME,
};

const CONFIG_OVERRIDES: Record<string, Partial<Config>> = {
    splitWithoutWrapping: {
        SCREEN_WIDTH: 80,
        MIN_LINE_WIDTH: 40,
        WRAP_LINES: false,
    },
    splitWithWrapping: {
        SCREEN_WIDTH: 80,
        MIN_LINE_WIDTH: 40,
        WRAP_LINES: true,
    },
    unifiedWithWrapping: {
        SCREEN_WIDTH: 80,
        MIN_LINE_WIDTH: 80,
        WRAP_LINES: true,
    },
    // This is in split mode
    inlineChangesHighlighted: {
        HIGHLIGHT_LINE_CHANGES: true,
        DELETED_WORD_COLOR: { color: { r: 255, g: 0, b: 0, a: 255 } },
        INSERTED_WORD_COLOR: { color: { r: 0, g: 255, b: 0, a: 255 } },
    },
    unifiedWithInlineChangesHighlighted: {
        SCREEN_WIDTH: 80,
        MIN_LINE_WIDTH: 80,
        HIGHLIGHT_LINE_CHANGES: true,
        DELETED_WORD_COLOR: { color: { r: 255, g: 0, b: 0, a: 255 } },
        INSERTED_WORD_COLOR: { color: { r: 0, g: 255, b: 0, a: 255 } },
    },
    syntaxHighlighted: {
        SCREEN_WIDTH: 80,
        MIN_LINE_WIDTH: 40,
        WRAP_LINES: false,
        SYNTAX_HIGHLIGHTING_THEME: 'dark-plus',
    },
};

for (const [configName, configOverride] of Object.entries(CONFIG_OVERRIDES)) {
    async function transform(input: string): Promise<string> {
        const testConfig: Config = {
            ...TEST_CONFIG,
            ...configOverride,
        };
        const context = await getContextForConfig(testConfig);

        let string = '';
        await transformContentsStreaming(
            context,
            Readable.from(input),
            new (class extends Writable {
                write(chunk: Buffer) {
                    string += chunk.toString();
                    return true;
                }
            })()
        );
        return string;
    }

    describe(configName, () => {
        test('empty', async function () {
            expect(await transform(``)).toMatchSnapshot();
        });

        test('with ANSI color codes', async function () {
            expect(
                await transform(`
[1;32mcommit f735de7025c6d626c5ae1a291fe24f143dea0313[m
Author: Shrey Banga <banga.shrey@gmail.com>
Date:   Sun Apr 11 15:25:34 2021 -0700

    Add theme support

[1;33mdiff --git a/todo.md b/todo.md[m
[1;33mindex 9f14e96..eaf3730 100644[m
[1;33m--- a/todo.md[m
[1;33m+++ b/todo.md[m
[1;32m@@ -7,6 +7,7 @@[m
 -   [x] Handle file addition/deletion properly[m
 -   [x] Fix incorrect line positions when a hunk has discontinuous inserts and/or deletes[m
 -   [x] Organize code[m
[1;32m+[m[1;32m-   [x] Move visual config to theme[m
 -   [ ] Handle empty diffs[m
 -   [ ] Handle moves and renames without diffs[m
 -   [ ] Highlight changes in lines[m

`)
            ).toMatchSnapshot();
        });

        test('commits without diffs', async function () {
            expect(
                await transform(`
commit e5f896655402f8cf2d947c528d45e1d56bbf5717 (HEAD -> main)
Author: Shrey Banga <banga.shrey@gmail.com>
Date:   Sun Apr 11 16:23:54 2021 -0700

    Small refactor to allow testing end-to-end

commit b637f38029f4a89c6a3b73b2b84a6a5b9e260730
Author: Shrey Banga <banga.shrey@gmail.com>
Date:   Sun Apr 11 11:53:02 2021 -0700

    Organize code

commit f323143e03af95fee5d38c21238a92ffd4461847
Author: Shrey Banga <banga.shrey@gmail.com>
Date:   Sun Apr 11 10:39:17 2021 -0700

    more todos
`)
            ).toMatchSnapshot();
        });

        test('commit with addition', async function () {
            expect(
                await transform(`
commit f735de7025c6d626c5ae1a291fe24f143dea0313
Author: Shrey Banga <banga.shrey@gmail.com>
Date:   Sun Apr 11 15:25:34 2021 -0700

    Add theme support

diff --git a/todo.md b/todo.md
index 9f14e96..eaf3730 100644
--- a/todo.md
+++ b/todo.md
@@ -9,2 +9,3 @@
 -   [x] Organize code
+-   [x] Move visual config to theme
 -   [ ] Handle empty diffs
`)
            ).toMatchSnapshot();
        });

        test('commit with deletion', async function () {
            expect(`commit eccfb5a2b3d76ba53df315f977da74b18d50113e
Author: Shrey Banga <shrey@quip.com>
Date:   Thu Aug 22 10:07:25 2019 -0700

    Remove deprecated option

diff --git a/Code/User/settings.json b/Code/User/settings.json
index a33d267..ae58a01 100644
--- a/Code/User/settings.json
+++ b/Code/User/settings.json
@@ -26,5 +26,4 @@
   // search
   "search.location": "panel",
-  "search.usePCRE2": true,

   // telemetry
`).toMatchSnapshot();
        });

        test('commit with a small diff', async function () {
            expect(