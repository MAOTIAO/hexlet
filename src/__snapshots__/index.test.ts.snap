// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inlineChangesHighlighted commit with a small diff 1`] = `
"
commit 26ca49fb83758bace20a473e231d576aa1bbe115                                                                         
Author: Shrey Banga <shrey@quip.com>                                                                                    
Date:   Tue May 23 16:47:17 2017 -0700                                                                                  
                                                                                                                        
    sonos to brew                                                                                                       
                                                                                                                        
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ Brewfile                                                                                                            
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@@ -19,2 +19,3 @@ brew 'python3'                                                                                        
   19    brew 'socat'                                          19    brew 'socat'                                       
                                                               20  + brew 'sonos'                                       
   20    brew 'terminal-notifier'                              21    brew 'terminal-notifier'                           
@@ -42,3 +43,2 @@ cask 'rescuetime'                                                                                     
   42    cask 'slate'                                          43    cask 'slate'                                       
   43  - cask 'sonos'                                                                                                   
   44    cask 'spotify'                                        44    cask 'spotify'                                     
"
`;

exports[`inlineChangesHighlighted commit with a small diff without a/ and b/ prefixs 1`] = `
"
commit 26ca49fb83758bace20a473e231d576aa1bbe115                                                                         
Author: Shrey Banga <shrey@quip.com>                                                                                    
Date:   Tue May 23 16:47:17 2017 -0700                                                                                  
                                                                                                                        
    sonos to brew                                                                                                       
                                                                                                                        
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ Brewfile                                                                                                            
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@@ -19,2 +19,3 @@ brew 'python3'                                                                                        
   19    brew 'socat'                                          19    brew 'socat'                                       
                                                               20  + brew 'sonos'                                       
   20    brew 'terminal-notifier'                              21    brew 'terminal-notifier'                           
@@ -42,3 +43,2 @@ cask 'rescuetime'                                                                                     
   42    cask 'slate'                                          43    cask 'slate'                                       
   43  - cask 'sonos'                                                                                                   
   44    cask 'spotify'                                        44    cask 'spotify'                                     
"
`;

exports[`inlineChangesHighlighted commit with addition 1`] = `
"
commit f735de7025c6d626c5ae1a291fe24f143dea0313                                                                         
Author: Shrey Banga <banga.shrey@gmail.com>                                                                             
Date:   Sun Apr 11 15:25:34 2021 -0700                                                                                  
                                                                                                                        
    Add theme support                                                                                                   
                                                                                                                        
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ todo.md                                                                                                             
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@@ -9,2 +9,3 @@                                                                                                         
    9    -   [x] Organize code                                  9    -   [x] Organize code                              
                                                               10  + -   [x] Move visual config to theme                
   10    -   [ ] Handle empty diffs                            11    -   [ ] Handle empty diffs                         
   11                                                          12                                                       
"
`;

exports[`inlineChangesHighlighted commit with binary file addition 1`] = `
"
commit c0ca4394fd55f1709430414f03db3d04cb9cc72c (HEAD -> main)                                                          
Author: Shrey Banga <banga.shrey@gmail.com>                                                                             
Date:   Wed Apr 14 18:24:29 2021 -0700                                                                                  
                                                                                                                        
    test binary file addition                                                                                           
                                                                                                                        
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ screenshots/default.png                                                                                             
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ test dir a/default.png                                                                                              
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
"
`;

exports[`inlineChangesHighlighted commit with deletion 1`] = `
"commit eccfb5a2b3d76ba53df315f977da74b18d50113e
Author: Shrey Banga <shrey@quip.com>
Date:   Thu Aug 22 10:07:25 2019 -0700

    Remove deprecated option

diff --git a/Code/User/settings.json b/Code/User/settings.json
index a33d267..ae58a01 100644
--- a/Code/User/settings.json
+++ b/Code/User/settings.json
@@ -26,5 +26,4 @@
   // search
   \\"search.location\\": \\"panel\\",
-  \\"search.usePCRE2\\": true,

   // telemetry
"
`;

exports[`inlineChangesHighlighted commit with file addition 1`] = `
"
commit e4951eee3b9a8fa471d01dd64075c5fd44879a26                                                                         
Author: Shrey Banga <banga.shrey@gmail.com>                                                                             
Date:   Sat Apr 10 14:35:42 2021 -0700                                                                                  
                                                                                                                        
    wip                                                                                                                 
                                                                                                                        
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ .gitignore                                                                                                          
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@@ -0,0 +1,2 @@                                                                                                         
                                                                1  + node_modules/**                                    
                                                                2  + build/**                                           
                                                                3    No newline at end of file                          
"
`;

exports[`inlineChangesHighlighted commit with file deletion 1`] = `
"
commit 1e6ebaccc6fadf3390a749e7aa2cc6372b24325e                                                                         
Author: Shrey Banga <banga.shrey@gmail.com>                                                                             
Date:   Sun Apr 18 03:17:05 2021 -0700                                                                                  
                                                                                                                        
    Update main.yml                                                                                                     
                                                                                                                        
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■                                                                                                                     
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ scripts/screenshotTheme                                                                                             
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@@ -1,7 +0,0 @@                                                                                                         
    1  - on run argv                                                                                                    
    2  -     with timeout of 5 seconds                                                                                  
    3  -         tell app \\"Terminal\\"                                                                                    
    4  -             open \\"screenshots.sh\\"                                                                              
    5  -         end tell                                                                                               
    6  -     end timeout                                                                                                
    7  - end run                                                                                                        
    8    No newline at end of file                                                                                      
"
`;

exports[`inlineChangesHighlighted commit with file move 1`] = `
"
commit 1c76ed4bb05429741fd4a48896bb84b11bc661f5                                                                         
Author: Shrey Banga <banga.shrey@gmail.com>                                                                             
Date:   Sat Apr 10 22:26:15 2021 -0700                                                                                  
                                                                                                                        
    Move sample diff files                                                                                              
                                                                                                                        
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ colors.diff -> test-data/colors.diff                                                                                
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
"
`;

exports[`inlineChangesHighlighted commit with inline line changes 1`] = `
"
commit 9e424d329800e945e3003c9e275f80bdef69e591                                                                         
Author: Shrey Banga <banga.shrey@gmail.com>                                                                             
Date:   Sun Apr 4 19:04:26 2021 -0700                                                                                   
                                                                                                                        
    update lockfile                                                                                                     
                                                                                                                        
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ Brewfile.lock.json                                                                                                  
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@@ -5 +5 @@                                                                                                             
    5  -         \\"revision\\": ░░░░                               5  +         \\"revision\\": ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
@@ -8 +8 @@                                                                                                             
    8  -         \\"revision\\": ░░░░                               8  +         \\"revision\\": ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
@@ -11 +11 @@                                                                                                           
   11  -         \\"revision\\": ░░░░                              11  +         \\"revision\\": ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
@@ -16 +16 @@                                                                                                           
   16  -         \\"version\\": \\"1.░░.░░░\\",                        16  +         \\"version\\": \\"1.░░.░\\",                       
@@ -17,0 +18 @@                                                                                                         
                                                               18  +           \\"rebuild\\": 0,                            
@@ -19,0 +21 @@                                                                                                         
                                                               21  +           \\"root_url\\": \\"https://homebrew.bintray.com
@@ -20,0 +23,8 @@                                                                                                       
                                                               23  +             \\"arm64_big_sur\\": {                     
                                                               24  +               \\"url\\": \\"https://homebrew.bintray.com/
                                                               25  +               \\"sha256\\": \\"e7c1efdd09e951eb46d01a3200
                                                               26  +             },                                     
                                                               27  +             \\"big_sur\\": {                           
                                                               28  +               \\"url\\": \\"https://homebrew.bintray.com/
                                                               29  +               \\"sha256\\": \\"ea37f33fd27369612a3e4e6db6
                                                               30  +             },                                     
@@ -22,2 +32,2 @@                                                                                                       
   22  -               \\"url\\": \\"https://homebrew.bintray.com/   32  +               \\"url\\": \\"https://homebrew.bintray.com/
   23  -               \\"sha256\\": \\"░░░░░░░░░░░░░░░░░░░░░░░░░░   33  +               \\"sha256\\": \\"░░░░░░░░░░░░░░░░░░░░░░░░░░
@@ -26,6 +36,2 @@                                                                                                       
   26  -               \\"url\\": \\"https://homebrew.bintray.com/   36  +               \\"url\\": \\"https://homebrew.bintray.com/
   27  -               \\"sha256\\": \\"░░░░░░░░░░░░░░░░░░░░░░░░░░   37  +               \\"sha256\\": \\"░░░░░░░░░░░░░░░░░░░░░░░░░░
   28  -             },                                                                                                 
   29  -             \\"high_sierra\\": {                                                                                   
   30  -               \\"url\\": \\"https://homebrew.bintray.com/                                                            
   31  -               \\"sha256\\": \\"0997f6f5cda0e3bdb7789a80b5                                                            
   32                                                          38                                                       
"
`;

exports[`inlineChangesHighlighted commit with tabs 1`] = `
"
commit f735de7025c6d626c5ae1a291fe24f143dea0313                                                                         
Author: Shrey Banga <banga.shrey@gmail.com>                                                                             
Date:   Sun Apr 11 15:25:34 2021 -0700                                                                                  
                                                                                                                        
    Add theme support                                                                                                   
                                                                                                                        
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ■■ todo.md                                                                                                             
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@@ -9,2 +9,3 @@                                                                                                         
    9    -        [x] Organize code                             9    -        [x] Organize code                         
                                                               10  + -        [x] Move visual config to theme           
   10    -        [ ] Handle empty diffs                       11    -        [ ] Handle empty diffs                    
   11                                                          12                                                       
"
`;

exports[`inlineChangesHighlighted commits without diffs 1`] = `
"
commit e5f896655402f8cf2d947c528d45e1d56bbf5717 (HEAD -> main)                                                          
Author: Shrey Banga <banga.shrey@gmail.com>                                                                             
Date:   Sun Apr 11 16:23:54 2021 -0700                                                                                  
                                                                                                                        
    Small refactor to allow testing end-to-end                                                                          
                                                                                                                        
commit b637f38029f4a89c6a3b73b2b84a6a5b9e260730                                                                         
Author: Shrey Banga <banga.shrey@gmail.com>                                                                             
Date:   Sun Apr 11 11:53:02 2021 -0700                                                                                  
                                                                                                                        
    Organize code                                                                                                       
                                                    